src/actions/mod.rs

```diff
@@ -197,21 +197,81 @@ impl InitActionContext {
     }
 
     fn build<O: Output>(&self, project_path: &Path, priority: BuildPriority, out: O) {
-        struct BuildNotifier<O: Output> {
+
+        lazy_static! {
+            static ref PROGRESS_ID_COUNTER: AtomicUsize = {
+                AtomicUsize::new(0)
+            };
+        }
+
+        fn new_progress_params(id: String, title: &str) -> ProgressParams {
+            ProgressParams {
+                id,
+                title: Some(title.into()),
+                message: None,
+                percentage: None,
+                done: None,
+            }
+        }
+
+        // notifier of diagnostics after the build
+        struct BuildDiagnosticsNotifier<O: Output> {
             out: O,
             active_build_count: Arc<AtomicUsize>,
+            progress_params: ProgressParams,
         }
 
-        impl<O: Output> Notifier for BuildNotifier<O> {
-            fn notify_begin(&self) {
-                self.out.notify(Notification::<DiagnosticsBegin>::new(NoParams {}));
+        // base progress parameters for notification of the analysis.
+        let diganostics_progress_params = {
+            let id = format!("diagnostics_{}", PROGRESS_ID_COUNTER.fetch_add(1, Ordering::SeqCst));
+            new_progress_params(id, "Diagnostics")
+        };
+
+        impl<O: Output> DiagnosticsNotifier for BuildDiagnosticsNotifier<O> {
+            fn notify_begin_diagnostics(&self) {
+                let params = self.progress_params.clone();
+                self.out.notify(Notification::<Progress>::new(params));
+            }
+            fn notify_publish_diagnostics(&self, params: PublishDiagnosticsParams) {
+                self.out.notify(Notification::<PublishDiagnostics>::new(params));
             }
-            fn notify_end(&self) {
+            fn notify_end_diagnostics(&self) {
                 self.active_build_count.fetch_sub(1, Ordering::SeqCst);
-                self.out.notify(Notification::<DiagnosticsEnd>::new(NoParams {}));
+                let mut params = self.progress_params.clone();
+                params.done = Some(true);
+                self.out.notify(Notification::<Progress>::new(params));
             }
-            fn notify_publish(&self, params: PublishDiagnosticsParams) {
-                self.out.notify(Notification::<PublishDiagnostics>::new(params));
+        }
+
+        // notifier of progress for the build (window/progress notifications)
+        struct BuildProgressNotifier<O: Output> {
```

Comment:
Similarly to `BuildDiagnosticsNotifier`, this would probably benefit from moving it together to a separate place